### ISSUE TITLE
Error occured during transcript mode running

### ISSUE THREAD

**Issue by aman21392** [bcgsc/NanoSim#124]
Hi,
I using the transcript mode command as below and gives an error, please can you tell me what's wrong.

 nohup /home/aclab/apps/NanoSim/src/read_analysis.py transcriptome -i control_T.fastq -rg Homo_sapiens.GRCh38.dna.primary_assembly.fa -rt Homo_sapiens.GRCh38.all.ncrna_cdna.fa -annot Homo_sapiens.GRCh38.104.chr.gff3.gz -o ./transcript/control_2nd -t 70 &

running the code with following parameters:

infile /Drive7/2nd_nanopore_experiment_data/control_25march/20200316_1633_MN33429_FAM96501_e0563569/fastq_pass/control_T.fastq
ref_g /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa
ref_t /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.all.ncrna_cdna.fa
annot /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.104.chr.gff3.gz
aligner minimap2
g_alnm 
t_alnm 
prefix /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd
num_threads 70
model_fit True
intron_retention True
2021-06-07 16:40:38: Read pre-process and unaligned reads analysis
2021-06-07 16:41:34: Alignment with minimap2 to reference transcriptome
[M::mm_idx_gen::8.256*1.62] collected minimizers
[M::mm_idx_gen::9.141*3.14] sorted minimizers
[M::main::9.207*3.13] loaded/built the index for 257575 target sequence(s)
[M::mm_mapopt_update::9.769*3.00] mid_occ = 144
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 257575
[M::mm_idx_stat::9.998*2.96] distinct minimizers: 20086155 (40.89% are singletons); average occurrences: 3.899; average spacing: 5.403; total length: 423105511
[M::worker_pipeline::37.708*4.26] mapped 834111 sequences
[M::worker_pipeline::44.025*3.68] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax map-ont -t 70 /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.all.ncrna_cdna.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd_processed.fasta.gz
[M::main] Real time: 44.278 sec; CPU: 162.340 sec; Peak RSS: 4.324 GB
2021-06-07 16:42:19: Alignment with minimap2 to reference genome
[M::mm_idx_gen::59.740*1.94] collected minimizers
[M::mm_idx_gen::70.295*5.54] sorted minimizers
[M::main::70.295*5.54] loaded/built the index for 194 target sequence(s)
[M::mm_mapopt_update::73.690*5.33] mid_occ = 765
[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 194
[M::mm_idx_stat::75.802*5.21] distinct minimizers: 167225302 (35.46% are singletons); average occurrences: 6.030; average spacing: 3.074; total length: 3099750718
[M::worker_pipeline::141.134*18.16] mapped 834111 sequences
[M::worker_pipeline::150.662*17.04] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax splice -t 70 /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd_processed.fasta.gz
[M::main] Real time: 151.051 sec; CPU: 2567.665 sec; Peak RSS: 31.310 GB
2021-06-07 16:44:51: Processing transcriptome alignment file: bam
2021-06-07 16:44:59: Processing genome alignment file: bam
2021-06-07 16:45:18: Parse the annotation file (GTF/GFF3)
/bin/sh: 1: gt: not found
/bin/sh: 1: cannot open /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd_added_intron_temp.gff3: No such file
/bin/sh: 1: gt: not found
rm: cannot remove '/Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd_added_intron_temp.gff3': No such file or directory
2021-06-07 16:45:18: Modeling Intron Retention
2021-06-07 16:45:18: Reading intron coordinates from GFF file
Traceback (most recent call last):
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 720, in <module>
    main()
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 675, in main
    model_ir.intron_retention(prefix, prefix + "_added_intron_final.gff3", g_alnm, t_alnm)
  File "/home/aclab/apps/NanoSim/src/model_intron_retention.py", line 40, in intron_retention
    for feature in gff_features:
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/features.py", line 133, in __iter__
    for line in super().__iter__():
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/utils.py", line 53, in __iter__
    self.__enter__()
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/utils.py", line 38, in __enter__
    lines = open(self.fos)
FileNotFoundError: [Errno 2] No such file or directory: '/Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_2nd_added_intron_final.gff3'

Thanks in advance

**Comment by SaberHQ** [comment]
Hey @aman21392 

I believe this is due to the **_.gz_** format of the GFF file. Unfortunately, NanoSim does not accept it and it causes the error you got as the pipeline couldn't handle and process the GFF file inside the **_add_intron_** function. 

We can definitely take a look at it and add that feature to the NanoSim pipeline very shortly. Meanwhile, would you please unzip your GFF file and provide it as input to the **_read_analysis.py_** script and let me know if it works? So the name of the input annot file will be something like: **_Homo_sapiens.GRCh38.104.chr.gff3_**

I will keep you updated whenever we implement the support for .gz format for annotation files. Thanks for your interest in our tool and please let me know if you have any more questions.

**Comment by aman21392** [comment]
Hi saber,
I used below command-
/home/aclab/apps/NanoSim/src/read_analysis.py transcriptome -i control_T.fastq -rg Homo_sapiens.GRCh38.dna.primary_assembly.fa -rt Homo_sapiens.GRCh38.all.ncrna_cdna.fa -annot Homo_sapiens.GRCh38.104.chr.gff3 -o transcript/control_intron -t 70 &

Again it gives the same error, please look at it and sort this issue.
########################################################################
running the code with following parameters:

infile /Drive7/2nd_nanopore_experiment_data/control_25march/20200316_1633_MN33429_FAM96501_e0563569/fastq_pass/control_T.fastq
ref_g /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa
ref_t /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.all.ncrna_cdna.fa
annot /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.104.chr.gff3
aligner minimap2
g_alnm 
t_alnm 
prefix /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron
num_threads 70
model_fit True
intron_retention True
2021-06-09 13:02:24: Read pre-process and unaligned reads analysis
2021-06-09 13:03:21: Alignment with minimap2 to reference transcriptome
[M::mm_idx_gen::7.727*1.60] collected minimizers
[M::mm_idx_gen::8.284*3.22] sorted minimizers
[M::main::8.343*3.20] loaded/built the index for 257575 target sequence(s)
[M::mm_mapopt_update::8.902*3.06] mid_occ = 144
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 257575
[M::mm_idx_stat::9.129*3.01] distinct minimizers: 20086155 (40.89% are singletons); average occurrences: 3.899; average spacing: 5.403; total length: 423105511
[M::worker_pipeline::36.564*4.41] mapped 834111 sequences
[M::worker_pipeline::42.726*3.80] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax map-ont -t 70 /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/check/Homo_sapiens.GRCh38.all.ncrna_cdna.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_processed.fasta.gz
[M::main] Real time: 42.861 sec; CPU: 162.599 sec; Peak RSS: 3.687 GB
2021-06-09 13:04:04: Alignment with minimap2 to reference genome
[M::mm_idx_gen::61.654*1.93] collected minimizers
[M::mm_idx_gen::71.571*5.30] sorted minimizers
[M::main::71.572*5.30] loaded/built the index for 194 target sequence(s)
[M::mm_mapopt_update::75.612*5.07] mid_occ = 765
[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 194
[M::mm_idx_stat::77.685*4.96] distinct minimizers: 167225302 (35.46% are singletons); average occurrences: 6.030; average spacing: 3.074; total length: 3099750718
[M::worker_pipeline::136.826*16.01] mapped 834111 sequences
[M::worker_pipeline::146.618*14.97] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax splice -t 70 /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_processed.fasta.gz
[M::main] Real time: 146.948 sec; CPU: 2194.662 sec; Peak RSS: 31.275 GB
2021-06-09 13:06:32: Processing transcriptome alignment file: bam
2021-06-09 13:06:40: Processing genome alignment file: bam
2021-06-09 13:06:59: Parse the annotation file (GTF/GFF3)
/bin/sh: 1: gt: not found
/bin/sh: 1: cannot open /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_added_intron_temp.gff3: No such file
/bin/sh: 1: gt: not found
rm: cannot remove '/Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_added_intron_temp.gff3': No such file or directory
2021-06-09 13:06:59: Modeling Intron Retention
2021-06-09 13:06:59: Reading intron coordinates from GFF file
Traceback (most recent call last):
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 720, in <module>
    main()
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 675, in main
    model_ir.intron_retention(prefix, prefix + "_added_intron_final.gff3", g_alnm, t_alnm)
  File "/home/aclab/apps/NanoSim/src/model_intron_retention.py", line 40, in intron_retention
    for feature in gff_features:
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/features.py", line 133, in __iter__
    for line in super().__iter__():
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/utils.py", line 53, in __iter__
    self.__enter__()
  File "/usr/local/lib/python3.6/dist-packages/HTSeq/utils.py", line 38, in __enter__
    lines = open(self.fos)
FileNotFoundError: [Errno 2] No such file or directory: '/Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_added_intron_final.gff3'
########################################################################
The files making are-
-rw-rw-r-- 1 aclab aclab 179M Jun  9 13:03 control_intron_processed.fasta.gz
-rw-rw-r-- 1 aclab aclab 237M Jun  9 13:04 control_intron_transcriptome_alnm.bam
-rw-rw-r-- 1 aclab aclab 330M Jun  9 13:06 control_intron_genome_alnm.bam
-rw-rw-r-- 1 aclab aclab  26M Jun  9 13:06 control_intron_transcriptome_primary.bam
-rw-rw-r-- 1 aclab aclab 103M Jun  9 13:06 control_intron_genome_primary.bam
-rw-rw-r-- 1 aclab aclab 473M Jun  9 13:06 control_intron.gff3
########################################################################
Thanks in advance

**Comment by kmnip** [comment]
This is the relevant part of the error log:

> /bin/sh: 1: gt: not found

It couldn't find the program `gt`, which is GenomeTools.

@cheny19 , It doesn't look like GenomeTools was specified in the required packages and the conda recipe.

**Comment by cheny19** [comment]
@kmnip , thanks for pointing it out. Unfortunately, GeomeTools is not supported on Conda any more. So @aman21392 you'll have to install it following their [manual](http://genometools.org) to use GenomeTools. I'll also mention that in the README.

**Comment by aman21392** [comment]
Hi,
There is another error that occurred after solving the genome tools error. The below command used it :
nohup /home/aclab/apps/NanoSim/src/read_analysis.py transcriptome -i control_T.fastq -rg Homo_sapiens.GRCh38.dna.primary_assembly.fa -rt Homo_sapiens.GRCh38.all.ncrna_cdna.fa -annot Homo_sapiens.GRCh38.104.chr.gff3 -o ./transcript/control_intron -t 70
 I searched about below error and it says you used .gz file  but in command you also see that there is no .gz file I used it.

running the code with following parameters:

infile /Drive7/2nd_nanopore_experiment_data/control_25march/20200316_1633_MN33429_FAM96501_e0563569/fastq_pass/control_T.fastq
ref_g /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa
ref_t /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/latest_cDNA/Homo_sapiens.GRCh38.all.ncrna_cdna.fa
annot /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/latest_cDNA/Homo_sapiens.GRCh38.104.chr.gff3
aligner minimap2
g_alnm 
t_alnm 
prefix /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron
num_threads 70
model_fit True
intron_retention True
2021-06-11 20:54:30: Read pre-process and unaligned reads analysis
2021-06-11 20:55:29: Alignment with minimap2 to reference transcriptome
[M::mm_idx_gen::7.822*1.58] collected minimizers
[M::mm_idx_gen::8.380*2.97] sorted minimizers
[M::main::8.436*2.96] loaded/built the index for 257575 target sequence(s)
[M::mm_mapopt_update::8.971*2.84] mid_occ = 144
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 257575
[M::mm_idx_stat::9.205*2.79] distinct minimizers: 20086155 (40.89% are singletons); average occurrences: 3.899; average spacing: 5.403; total length: 423105511
[M::worker_pipeline::37.168*4.33] mapped 834111 sequences
[M::worker_pipeline::43.576*3.73] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax map-ont -t 70 /Drive4/nanopore_2nd_experiment/ncRNA_cdna/cDNA/latest_cDNA/Homo_sapiens.GRCh38.all.ncrna_cdna.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_processed.fasta.gz
[M::main] Real time: 43.817 sec; CPU: 162.743 sec; Peak RSS: 4.332 GB
2021-06-11 20:56:13: Alignment with minimap2 to reference genome
[M::mm_idx_gen::58.983*1.95] collected minimizers
[M::mm_idx_gen::66.971*4.98] sorted minimizers
[M::main::66.971*4.98] loaded/built the index for 194 target sequence(s)
[M::mm_mapopt_update::70.009*4.81] mid_occ = 765
[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 194
[M::mm_idx_stat::72.215*4.69] distinct minimizers: 167225302 (35.46% are singletons); average occurrences: 6.030; average spacing: 3.074; total length: 3099750718
[M::worker_pipeline::131.832*16.26] mapped 834111 sequences
[M::worker_pipeline::141.601*15.17] mapped 258720 sequences
[M::main] Version: 2.17-r974-dirty
[M::main] CMD: minimap2 --cs -ax splice -t 70 /Drive4/nanopore_2nd_experiment/human_genome/Homo_sapiens.GRCh38.dna.primary_assembly.fa /Drive7/2nd_nanopore_experiment_data/control_25march/nanosim/2nd_try_characterize_intron_with/transcript/control_intron_processed.fasta.gz
[M::main] Real time: 141.958 sec; CPU: 2148.407 sec; Peak RSS: 30.940 GB
2021-06-11 20:58:36: Processing transcriptome alignment file: bam
2021-06-11 20:58:44: Processing genome alignment file: bam
2021-06-11 20:59:04: Parse the annotation file (GTF/GFF3)
2021-06-11 21:00:00: Modeling Intron Retention
2021-06-11 21:00:00: Reading intron coordinates from GFF file
2021-06-11 21:01:19: Read primary genome alignment for each read
Traceback (most recent call last):
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 720, in <module>
    main()
  File "/home/aclab/apps/NanoSim/src/read_analysis.py", line 675, in main
    model_ir.intron_retention(prefix, prefix + "_added_intron_final.gff3", g_alnm, t_alnm)
  File "/home/aclab/apps/NanoSim/src/model_intron_retention.py", line 72, in intron_retention
    for alnm in g_alignments:
  File "/home/aclab/.local/lib/python3.6/site-packages/HTSeq/__init__.py", line 574, in __iter__
    for line in FileOrSequence.__iter__(self):
  File "/home/aclab/.local/lib/python3.6/site-packages/HTSeq/__init__.py", line 61, in __iter__
    for line in lines:
  File "/usr/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte

Please can you sort out this issue?
Thanks in advance

**Comment by SaberHQ** [comment]
I was able to replicate the error @aman21392 mentioned above. 

I agree with you @aman21392. Parsing the annotation file works fine and there is nothing wrong about the gff file you used. The error happens in "model_intron_retention.py" where it is reading the genome alignment file using HTSeq SAM Reader whereas the format is bam.

To conclude, to make it simple, I should say that we added support for BAM files recently and we need to update it for IR modeling feature to support BAM formats as well. I am pinging my colleague @kmnip for his thoughts on this. 

I will keep you updated as soon as we fix this issue. Thanks for your patience. Cheers.
